analysis/filter/StructureModifier/src/main/kotlin/de/maibornwolff/codecharta/filter/structuremodifier/NodeRemover.kt

```diff
@@ -0,0 +1,52 @@
+package de.maibornwolff.codecharta.filter.structuremodifier
+
+import de.maibornwolff.codecharta.model.*
+
+class NodeRemover(private val project: Project, private val projectName: String?) {
+
+    fun remove(paths: Array<String>): Project {
+        val pathSegments = paths.map { it.removePrefix("/").removeSuffix("/").split("/") }
+        return ProjectBuilder(
+                projectName ?: project.projectName,
+                removeNodes(pathSegments),
+                removeEdges(paths),
+                copyAttributeTypes(),
+                removeBlacklistItems(paths)
+        ).build()
+    }
+
+    private fun filterNodes(path: List<String>, node: MutableNode): MutableNode {
+        val filteredChildren = node.children.filter { it.name != path.firstOrNull() || path.size > 1 }
+        node.children = filteredChildren.map {
+            filterNodes(path.drop(1), it)
+        }.toMutableList()
+        return node
+    }
+
+    private fun removeNodes(paths: List<List<String>>): MutableList<MutableNode> {
+        var root = project.rootNode.toMutableNode()
+        paths.forEach { root = filterNodes(it.drop(1), root) }
+        return mutableListOf(root)
+    }
+
+    private fun removeEdges(removePatterns: Array<String>): MutableList<Edge> {
+        var edges = project.edges
+        removePatterns.forEach { path -> edges = edges.filter { !it.fromNodeName.contains(path) && !it.toNodeName.contains(path) } }
```

Comment:
Actually the `edge.fromNodeName` and `edge.toNodeName` is possibly a gitignore pattern, so matching the path-string doesn't cover all options. Also see line 48 in `removeBlacklistItems`

Reply:
The more appropriate approach for the `NodeRemover` were the following simplified code. `newGitignoreList` in this case can be either edge-paths or the blacklist-paths.
```
newGitignoreList = gitignorePathList.filter { gitignorePathMatchesAnyNode(it, newRootNode) }
```

I admit its still very tricky for the `SubProjectExtractor` and `FolderMover`...

Reply:
do you know any java/kotlin library that supports gitignore pathmatching? I was searching for this in another context and did not find anything promising then

Reply:
Unfortunately not.

Reply:
In which case is the edge a gitignore pattern?

Reply:
True, the edge is currently never created as a gitignore pattern, but the blacklisted nodes can be

Reply:
In this case I would suggest we leave it for now, since potentially irrelevant gitignore patterns can be removed in the visualization afterwards. I think implementing a Gitignore matcher for that may be a bit of an overkill, since as a user I would not necessarily expect the blacklist entries to be checked for relevance if i remove some nodes. If we find a library for Gitignore matching we can add it later. What do you think?